From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 16 Jul 2023 13:03:27 +0200
Subject: [PATCH] Fiddle feature support flag

Adds a flag that indicates whether Fiddle features are supported in the current server execution.
Currently, this is true on any execution of the real server, but false on a dummy server started for the purpose of unit testing.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index dae36c6452ccd57a436dd918547b64d59957ab0a..80f587c1b16c59dfa724f9fdda847ffedb48d787 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -63,6 +63,7 @@ import net.minecraft.world.level.storage.LevelStorageSource;
 import net.minecraft.world.level.storage.LevelSummary;
 import net.minecraft.world.level.storage.PrimaryLevelData;
 import net.minecraft.world.level.storage.WorldData;
+import org.fiddlemc.fiddle.server.FiddleFeatureSupport;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -81,6 +82,7 @@ public class Main {
 
     @DontObfuscate
     public static void main(final OptionSet optionset) { // CraftBukkit - replaces main(String[] astring)
+        FiddleFeatureSupport.setSupportsFiddleFeatures(); // Fiddle - runtime environment feature support flag
         SharedConstants.tryDetectVersion();
         /* CraftBukkit start - Replace everything
         OptionParser optionparser = new OptionParser();
diff --git a/src/main/java/org/fiddlemc/fiddle/server/FiddleFeatureSupport.java b/src/main/java/org/fiddlemc/fiddle/server/FiddleFeatureSupport.java
new file mode 100644
index 0000000000000000000000000000000000000000..288a0786b388ed93b4c152f77c661e96cd8e605c
--- /dev/null
+++ b/src/main/java/org/fiddlemc/fiddle/server/FiddleFeatureSupport.java
@@ -0,0 +1,28 @@
+// Fiddle - runtime environment feature support flag
+
+package org.fiddlemc.fiddle.server;
+
+/**
+ * A utility class providing the {@link #supportsFiddleFeatures()} method.
+ */
+public final class FiddleFeatureSupport {
+
+    private FiddleFeatureSupport() {
+        throw new RuntimeException();
+    }
+
+    private static boolean supportsFiddleFeatures = false;
+
+    /**
+     * @return Whether the current server run supports Fiddle features.
+     * Typically, this is true on the real server, but false on a dummy server started for the purpose of unit testing.
+     */
+    public static boolean supportsFiddleFeatures() {
+        return supportsFiddleFeatures;
+    }
+
+    public static void setSupportsFiddleFeatures() {
+        supportsFiddleFeatures = true;
+    }
+
+}
